Fixes #139: First Input is First Event with InteractionId

mmocny opened the pull request:
At first I tried to make the minimal change to just update First-Input to be in terms of any Event Timing that has an interactionId, such that we can re-use the logic of computing interactionId and needn't manage a separate list of pending pointer downs for first-input, etc.

However, on the way to doing so I found several gaps and errors and just started to shave the yak.  There are a few changes which might have been possible to split out, but will be needed to resolve other in-flight changes (like for #145, #146, #147, #149).

I decided to add more text, update examples, move a few parts of the spec around to places it made more sense.

(This patch also doesn't quite fix all the things wrong with the spec even with the existing features not listed in the bugs above).


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/mmocny/event-timing/pull/151.html" title="Last updated on Mar 27, 2025, 2:11 PM UTC (612c107)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/event-timing/151/e1f728a...mmocny:612c107.html" title="Last updated on Mar 27, 2025, 2:11 PM UTC (612c107)">Diff</a>

mmocny commented:
@canova Please take a look (I cannot actually mark you as a reviewer for the patch-- wonder if we need to add you to the organization somehow?)!

Cheers.

shaseley commented on index.bs:
```diff
@@ -209,14 +219,9 @@ but a developer can set up a {{PerformanceObserver}} to observe future entries w
-This enables computing the <b>first input delay</b>, the <b>delay</b> of the <b>first input</b>.
+The Event Timing API also exposes timing information about the <b>first input</b> of a {{Window}},
+defined as the first {{PerformanceEventTiming}} entry with a non-0 {{PerformanceEventTiming/interactionId}}.
+This enables computing the <b>first input delay</b> (i.e. the <b>input delay</b> of the first Event of the <b>first interaction</b>).
```
s/Event/{{Event}}

shaseley commented on index.bs:
```diff
@@ -614,27 +609,33 @@ Finalize event timing {#sec-fin-event-timing}
         1. Let |pointerId| be |event|'s {{PointerEvent/pointerId}}.
-        1. If |pendingPointerDowns|[|pointerId|] exists, append |pendingPointerDowns|[|pointerId|] to |relevantGlobal|'s <a>entries to be queued</a>.
+        1. If |pendingPointerDowns|[|pointerId|] exists:
+            1. Let |previousPointerDownEntry| be |pendingPointerDowns|[|pointerId|].
```
Does this not get an interaction ID like the pending keydown below? IIUC that means it isn't a first-input candidate?

shaseley reviewed (commented):
Mostly prose suggestions, and one potential spec bug.

canova commented on index.bs:
```diff
@@ -664,19 +667,11 @@ Dispatch pending Event Timing entries {#sec-dispatch-pending}
-                1. Let <var>newFirstInputDelayEntry</var> be a copy of <var>timingEntry</var>.
-                1. Set <var>newFirstInputDelayEntry</var>'s {{PerformanceEntry/entryType}} to "<code>first-input</code>".
-                1. <a>Queue the entry</a> <var>newFirstInputDelayEntry</var>.
+    1. If |window|'s <a>has dispatched input event</a> is false, and |timingEntry|'s {{PerformanceEventTiming/interactionId}} is not 0, run the following steps:
```
Nice! I tested this new behavior in Firefox and all of the tests pass successfully. Previously I was still keeping the old logic with lots of checks. This simplifies this logic quite a lot. This is my diff with lots of red: https://gist.github.com/canova/b0a12c135c7a2385f64f95cc93db0136

mmocny replied:
Very nice!

canova commented on index.bs:
```diff
@@ -508,6 +501,12 @@ Increasing interaction count {#sec-increasing-interaction-count}
+This allows the user agent to choose to eagerly assign a <a>user interaction value</a> (i.e. at pointerdown) and then discard it (i.e. after pointercancel), rather than to lazily compute it.
+
+A user agent may choose to increase it by a small random integer every time, or choose a constant.
+A user agent must not use a shared global <a>user interaction value</a>s for all {{Window|Windows}}, because this could introduce cross origin leaks.
```
nit: I know this is just moving things around, but while here:
```suggestion
A user agent must not use a shared global <a>user interaction value</a>s for all {{Window|Windows}}, because this could introduce cross-origin leaks.
```

canova commented on index.bs:
```diff
@@ -147,12 +147,22 @@ Finally, this specification allows developers to obtain detailed information abo
 
-The very first user interaction has a disproportionate impact on user experience, and is often disproportionately slow.
+A single user interaction is typically actually made up several physical input gestures, each of which might dispatch several events on the page, and
+each of those might trigger multiple event listeners and/or default actions.  For example, a single user "tap" interaction on a touchscreen device might
```
I would say the same. Probably we can just not say "event" instead of "gesture" here.

mmocny replied:
I tried to look for wording from UI Events and Pointer Events specs for existing guidance on how to refer to what the user is physically doing, as a distinct concept from what the UA is doing.

Having done that, I think that I misused the term "gesture" here.  In those specs, "gesture" is often used to mean a long-running sequence of multiple physical input events from the user, in other words, the same concept as "Interaction" here. (example below)

"Interaction" is often also used as a synonym for "Gesture"... but really "Interaction" is super overloaded and sometimes used to mean anything from physical input to event dispatched.

I did find the term ["hysteresis"](https://www.w3.org/TR/uievents/#hysteresis) and the phrase "temporal hysteresis" in the glossary (in reference to how it takes time and multiple physical inputs to e.g. double-click).

One other strategy those specs sometimes use is to just be very explicit, and where it matters, use the text "physical user event" or "physical user input", or the phrasing "the user has ...", to distinguish how the user is physically making several distinct actions which collectively make up a "gesture"/"interaction".

I'll update wording throughout.

---

For example in the [UI Events spec, in (and above) example 9](https://www.w3.org/TR/uievents/#example-6453f9ad)

> A wheel event transaction is series of wheel events that are associated with a single user gesture.
> If a series of wheel events targetted in a scrollable element start above a child element, later events for the same user gesture may occur over the child element.

Also for example in [Pointer Events spec](https://www.w3.org/TR/pointerevents/#the-touch-action-css-property), in the note at the bottom of that section:

> Some user agents implement complex gestures for behaviors that involve a series of separate discrete gestures, but which are all treated as part of a single continuous gesture

...at first I thought this helped make a distinction made between `separate discrete gestures` and `part of a single continuous gesture`, but this example still describes how multiple long-running gestures made up of many physical inputs (flinger fling / swipe) can be joined together or separated (i.e. single or separate "interaction").

canova reviewed (commented):
Thanks for the PR!

I tested this logic locally in Firefox and it passes the tests successfully (after fixing the dispatch logic issues).

As you noted, this PR alone doesn't fix the `first-input` logic completely. Since we delay adding `pointerdown` and `keydown` to the "entries to be queued" it will select the wrong event entry. But seems like it's going to be fixed after the dispatch logic changes.


> @canova Please take a look (I cannot actually mark you as a reviewer for the patch-- wonder if we need to add you to the organization somehow?)!

Hmm yeah, I think I might need to be a part of org or repo to be able to ask for a review. I don't know how organizational things work in this repo though. Do you think we can ask someone?

r+ for the first-input dispatch logic. But looks like @shaseley has some valid points for the first paragraph.

canova commented:
@mmocny btw, you should be able to request reviews from me now. The permission issue should be resolved.

mmocny commented on index.bs:
```diff
@@ -147,12 +147,22 @@ Finally, this specification allows developers to obtain detailed information abo
 It's slow because it's often blocked on JavaScript execution that is not properly split into chunks during page load.
 The latency of the website's response to the first user interaction can be considered a key responsiveness and loading metric.
-To that effect, this API surfaces all the timing information about this interaction, even when this interaction is not handled slowly.
+To that effect, this API always surfaces the timing information about this first interaction, even if this interaction is not actually slow.
```
I changed the whole Introduction section, reduced duplication of some of the text, and made a lot of it more explicit (as suggested here).

There is now a while dedicated section to First Input.  Will upload latest for second round of review, shortly.

mmocny commented on index.bs:
```diff
@@ -614,27 +609,33 @@ Finalize event timing {#sec-fin-event-timing}
+            1. Add |previousKeyDownEntry| to |window|'s <a>entries to be queued</a>.
         1. Set |pendingKeyDowns|[|code|] to |timingEntry|.
+
+    1. Otherwise:
+        1. Append |timingEntry| to |relevantGlobal|’s <a>entries to be queued</a>.
```
For now this is just a trivial refactor of what we had before.  It used to be:

- If not pointerdown or keydown
- If pointerdown
- Else (keydown)

Now it is:

- If pointerdown
- If keydown
- Else (default)

---

In future patch(s), I'll fix the other parts of the spec which we discussion in other issues.

mmocny commented:
@canova @shaseley I think I've addressed the comments from last review-- but made significant changes to the introduction sections.

I found that there was some needless duplication of wording, while also not being specific enough in some cases.  I created distinct sub-sections for topics like Interactions and First Input, which I think makes things a lot clearer.

---

(Some) Future Steps:

- Remove pendingKeydown altogether.
- Move the pendingPointerdown handling from "finalize" steps back to "compute interaction id" algorithm.
  - We don't really need to wait for the finalization of the timing of the next pointerdown to update the previous pointerdown... we can do it as soon as we know it happened.
  - Moving things around this way should actually simplify several sections by isolating interactionId wording.
- Change the finalize algorithm:
  - Once we remove special cases for keydown and pointerdown..
  - Right now we immediately add "entries to be queued", too early.
  - We aren't actually done "finalizing" the event timing until we have presentation time.  We should move those parts of the spec up here.
  - Even once done "finalizing" the timing, we aren't ready for "entries to be queued" until *all* preceding entries have both an end time and an interactionId assigned.
    -  In practice, we just need to wait for pending Pointerdown to resolve.
  - We likely need to add another queue to split things up.  Not sure yet if its necessary to explicitly spec queueing events "by animation frame" (i.e. can we have a single queue of events, or do we need a list of lists, which is what Chrome does)

mmocny commented:
@canova I Updated this PR as only fixing #139 since it just fixes the definition, without fixing the ordering problem.

canova commented on index.bs:
```diff
@@ -147,20 +151,63 @@ Finally, this specification allows developers to obtain detailed information abo
+- {{click}},
+- ...and potentially some Focus events, Input events, etc, in between.
+
+All these individual {{UIEvent}}s will each produce a distinct {{PerformanceEventTiming}} entry, which is useful for detailed timing.
```
Looking at the list above, I think `pointermove` and `touchmove` doesn't produce a distinct `PerformanceEventTiming` entry. [They are part of excluded events above.](https://w3c.github.io/event-timing/#sec-events-exposed:~:text=mousemove%2C%20pointermove%2C%20pointerrawupdate%2C%20touchmove%2C%20wheel%2C%20and%20drag%20are%20excluded) So this sentence is not 100% true.

mmocny replied:
Good point-- will update.

canova reviewed (approved):
Looks good to me with a small nit. Thanks for working on it!

And thanks for listing the future steps as well!
